File: mgmt-back/app/services/pubsub/rabbit.py
import aio_pika
import asyncio
from typing import Callable, Coroutine, Any
import json
from .channels import CHANNELS, Channels

def serialize_json(message: dict) -> bytes:
    return json.dumps(message).encode('utf-8')

def deserialize_json(body: bytes) -> dict:
    return json.loads(body.decode('utf-8'))

class RabbitPubSubService:
    def __init__(self, host: str, port: int):
        self.rabbitmq_url = f"amqp://guest:guest@{host}:{port}/"
        self.connection = None
        self.channel = None

    async def connect(self):
        """Connect to RabbitMQ and declare queues."""
        self.connection = await aio_pika.connect_robust(self.rabbitmq_url)
        self.channel = await self.connection.channel()
        for queue_name in CHANNELS:
            await self.channel.declare_queue(queue_name, durable=True)

    async def sub(self, queue_name: Channels, callback: Callable[[Any], Coroutine]):
        """Subscribe to a queue and process messages with the given async callback."""
        if not self.channel:
            await self.connect()
        assert self.channel is not None
        queue = await self.channel.declare_queue(queue_name, durable=True)
        
        async def on_message(message: aio_pika.abc.AbstractIncomingMessage):
            async with message.process():
                data = deserialize_json(message.body)
                await callback(data)

        await queue.consume(on_message)
        print(f"Subscribed to {queue_name}")

    async def pub(self, queue_name: Channels, message: dict):
        """Publish a message to a queue."""
        if not self.channel:
            await self.connect()
        assert self.channel is not None
        body = serialize_json(message)
        await self.channel.default_exchange.publish(
            aio_pika.Message(
                body=body,
                delivery_mode=aio_pika.DeliveryMode.PERSISTENT
            ),
            routing_key=queue_name
        )
        

File: backend/components/pubsub/rabbit.ts
import * as amqp from "amqplib";
import { CHANNELS, Channels } from "./channels.js";

function serializeJson(message: Record<string, any>): Buffer {
  return Buffer.from(JSON.stringify(message));
}

function deserializeJson(body: Buffer): Record<string, any> {
  return JSON.parse(body.toString());
}

export class RabbitPubSubService {
  private rabbitmqUrl: string;
  private connection: any = null;
  private channel: any = null;

  constructor(host: string, port: number, user: string, password: string) {
    this.rabbitmqUrl = `amqp://${user}:${password}@${host}:${port}/`;
  }

  async connect(retries: number = 10, interval: number = 3000): Promise<void> {

    const maxRetries =
      parseInt(process.env.RABBITMQ_CONNECT_RETRIES || String(retries)) ||
      retries;
    const retryInterval =
      parseInt(process.env.RABBITMQ_CONNECT_INTERVAL || String(interval)) ||
      interval;

    let attempt = 0;
    let lastError: Error | null = null;

    while (true) {
      try {
        attempt++;
        this.connection = await amqp.connect(this.rabbitmqUrl);
        this.channel = await this.connection.createChannel();

        // Declare all queues as durable
        for (const queueName of CHANNELS) {
          await this.channel.assertQueue(queueName, { durable: true });
        }

        console.log("Connected to RabbitMQ");
        return;
      } catch (error) {
        lastError = error as Error;
        if (attempt >= maxRetries) {
          console.error(
            `Failed to connect to RabbitMQ after ${attempt} attempts:`,
            lastError
          );
          throw lastError;
        }
        console.log(
          `RabbitMQ not ready (attempt ${attempt}/${maxRetries}), retrying in ${retryInterval}ms...`
        );
        await new Promise((resolve) => setTimeout(resolve, retryInterval));
      }
    }
  }

  async pub(queueName: Channels, message: Record<string, any>): Promise<void> {
    /**
     * Publish a message to a queue.
     */
    if (!this.channel) {
      await this.connect();
    }

    if (!this.channel) {
      throw new Error("Channel is not initialized");
    }

    const body = serializeJson(message);

    await this.channel.sendToQueue(queueName, body, {
      persistent: true,
    });
  }

  async close(): Promise<void> {
    /**
     * Close the RabbitMQ connection gracefully.
     */
    if (this.channel) {
      await this.channel.close();
    }
    if (this.connection) {
      await this.connection.close();
    }
    console.log("RabbitMQ connection closed");
  }
}
